### PR TITLE
Set Truffle network config options for Truffle test's internal ganache network

### PR DESCRIPTION
**Pull request by sukanyaparashar**
### ISSUE
No way to set Truffle network config options for Truffle Test's internal Ganache network. See issue #4327 

### SOLUTION
This PR solves the issue of non-configurable Truffle test's managed ganache network from the truffle-config.js.
If the configured network is "test" and provider, host or url doesn't exist, or there is no configured network defined, then managed ganache is started overriding with the user specified configurations or the default configurations.
If there is a configured network in the config, then it connects to that external network.

### CODE REVIEW

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [inline packages/core/lib/commands/test/run.js:76]: This comment describes 4 conditions, but the code has three. I think the first conditional isn't in the implementation.

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [inline packages/core/lib/commands/test/run.js:83]: It would be nice to refactor for readability? Some conditional fragments may be useful further down the logic branches.
```javascript
  const configuredNetwork = config.networks[config.network];
  const testNetworkDefined = configuredNetwork && config.network == "test";
  const noProviderHostOrUrlConfigured = !configuredNetwork["provider"] && !configuredNetwork["host"] && !configuredNetwork["url"];

  if (testNetworkDefined && noProviderHostOrUrlConfigured) {
    // do the things
```

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [inline packages/core/lib/commands/test/run.js:86]: `port` is used in the other branches and could be factored out so it is only initialized once.

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [inline packages/core/lib/commands/test/run.js:139]: Maybe rename this function to `StartGanacheAndRunTests` is a better name for this function?

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [inline packages/core/lib/commands/test/run.js:105]: The definition of `numberOfFailures` could be hoisted above the if statement, and each branch would set `numberOfFailures` and there would be one return statement at the end.

**Review by cds-amal — CHANGES_REQUESTED** [review]
Thanks for this @sukanyaparashar. I noticed a discrepancy between the comment and code and made some refactoring/readability comments.

File: packages/core/lib/commands/test/run.js
Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:76]: The first and the last comment can be joined I guess, because that is the 2nd condition in the else if branch. I will change that. Thanks @cds-amal .

File: packages/core/lib/commands/test/run.js
Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:83]: Your suggestion in this code makes sense @cds-amal . I will make it more cleaner as suggested.

File: packages/core/lib/commands/test/run.js
Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:139]: Sure, this name will make more sense.

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [inline packages/core/lib/commands/test/run.js:82]: Oops, should verify `configuredNetwork` is not undefined! Sorry about that!
```suggestion
  const noProviderHostOrUrlConfigured =
    configuredNetwork &&
    !configuredNetwork.provider &&
    !configuredNetwork.host &&
    !configuredNetwork.url;
```

File: packages/core/lib/commands/test/run.js
Comment (haltman-at) [inline packages/core/lib/commands/test/run.js:88]: This condition fixes the behavior with `test`, but I don't think it preserves the existing behavior in other cases.  To preserve current behavior, it seems to me that this case case should also be invoked if `config.networks[config.netowork]` is undefined.  (This is the reason I hit "request changes".)

File: packages/core/lib/commands/test/run.js
Comment (haltman-at) [inline packages/core/lib/commands/test/run.js:84]: We probably shouldn't be getting a port unless we're actually in the managed case; I don't know how much work it actually takes to get a port, but it strikes me as probably something to be avoided in the cases where we don't need it.

File: packages/core/lib/commands/test/run.js
Comment (haltman-at) [inline packages/core/lib/commands/test/run.js:78]: I think this variable could use a better name; `testNetworkDefined` sounds like it refers to the `test` network, well, being defined, whereas what you're actually checking for is if it's defined *and* used.

**Review by haltman-at — DISMISSED** [review]
Seems mostly good, but there's one part that looks wrong to me; hitting "request changes" as a result.  Also two more minor points about variable naming & where code is placed.

File: packages/core/lib/commands/test/run.js
Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:88]: The last else branch actually checks for the undefined network. @cds-amal and I talked about this and I will make changes in the comments to make it more understandable.

File: packages/core/lib/commands/test/run.js
Comment (eggplantzzz) [reply packages/core/lib/commands/test/run.js:78]: I would say that generally in the Truffle codebase prefer `===` over `==` in most cases unless you actually want the behavior of `==`.

Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:72]: Sorry @eggplantzzz . I forgot about this again. I will correct it.

File: packages/core/lib/commands/test/run.js
Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:92]: Yes, @eggplantzzz . I think you are correct. We can just make it to if/else. I am working on it and testing it. I will make a PR with this change in a few minutes.

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [reply packages/core/lib/commands/test/run.js]: Nice rework on the refactoring @sukanyaparashar !

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [reply packages/core/lib/commands/test/run.js]: ooh, yea, that's a good point on the variable name

File: packages/core/lib/commands/test/run.js
Comment (cds-amal) [inline packages/core/lib/commands/test/run.js:116]: Perhaps leave a comment to indicate `configuredNetwork` is will only spread if it's defined and be ignored otherwise.

Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:116]: Sure, I will do that @cds-amal . Thanks.

File: packages/core/lib/commands/test/run.js
Comment (haltman-at) [inline packages/core/lib/commands/test/run.js:90]: This variable could still use a renaming like I suggested, I think.

File: packages/core/lib/commands/test/run.js
Comment (haltman-at) [inline packages/core/lib/commands/test/run.js:115]: You only use `port` in this block, so rather than declaring it with `let` up above, I think it would make more sense to declare it `const` right here.

**Review by haltman-at — COMMENTED** [review]
Cool, logic seems good now.  I'm dismissing my request changes.  I like the reorganization, too. :)  Two comments still on style though, hope you don't mind!

**Review by cds-amal — COMMENTED** [review]
This is shaping up nicely!

File: packages/core/lib/commands/test/run.js
Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:90]: Can we use "testNetworkDefinedAndUsed" as the variable name, @haltman-at ? Also suggested by @cds-amal . As it is related to the "test" network, so I am also in support of that. Any other suggestions anyone @eggplantzzz , @gnidan , @fainashalts ?

Comment (haltman-at) [reply packages/core/lib/commands/test/run.js:90]: Sure, that sounds fine.

Comment (sukanyaparashar) [reply packages/core/lib/commands/test/run.js:90]: Okay, I will change the it and push another commit. Thanks @haltman-at .

**Review by haltman-at — APPROVED** [review]
Seems good to me!

**Comment by sukanyaparashar** [comment]
> Seems good to me! Now we just have to see if @cds-amal still wants changes. :)

Sure, I will wait for @cds-amal to review it before merging.

**Review by cds-amal — APPROVED** [review]
LGTM! Nice work on this @sukanyaparashar .

**Comment by Miladinho** [comment]
Have the docs been updated to reflect this change? I tries to follow the PR description and have had no luck

**Pull request merged**
